### ISSUE TITLE
[Bug] Tachiyomi (Preview) crashes when loading a new chapter

### ISSUE THREAD

**Issue by Velorexe** [tachiyomiorg/tachiyomi#4982]
## Device information
* Tachiyomi version: r3328 (6aff438a)
* Android version: 9
* Device: G8341 (Sony Xperia XZ1)

## Steps to reproduce
1. Load manga from any source
2. Read until the chapters switch
3. Swipe to the first page of new chapter

### Expected behavior
Load in the new pages for the chapter and be able to read them.

### Actual behavior
Chance that Tachiyomi crashes. Restarting sometimes helps, other times it'll continue to show the same behavior.

## Other details
When swiping fast enough, you can catch a glimpse of the new chapter's first page, but app crashes after that.

### Crash Log
```--------- beginning of crash
04-29 11:17:14.431  4992  4992 E AndroidRuntime: FATAL EXCEPTION: main
04-29 11:17:14.431  4992  4992 E AndroidRuntime: Process: eu.kanade.tachiyomi.debug, PID: 4992
04-29 11:17:14.431  4992  4992 E AndroidRuntime: java.lang.ArrayIndexOutOfBoundsException: length=31; index=-1
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:439)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewerAdapter.onPageSplit(PagerViewerAdapter.kt:141)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewer$onPageSplit$1.run(PagerViewer.kt:390)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:280)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6706)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
--------- beginning of main
04-29 11:31:15.015  8175  8216 E SQLiteLog: (284) automatic index on C(manga_id)
04-29 11:31:15.015  8175  8216 E SQLiteLog: (284) automatic index on mangas_categories(manga_id)
04-29 11:31:18.551  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:18.618  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:19.141  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:19.225  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:20.788  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:20.821  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:21.142  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:21.183  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.605  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.617  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.626  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.644  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.664  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:22.690  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:23.416  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:23.447  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:23.869  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:23.900  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:24.599  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:24.688  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:24.936  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:25.034  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:31:27.041  8175  8183 E tachiyomi.debu: Something went wrong getting fds from adb. Retry!: Success
04-29 11:34:13.173  8175  8183 E tachiyomi.debu: Something went wrong getting fds from adb. Retry!: Connection refused
04-29 11:34:21.138  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:21.141  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:28.033  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:28.685  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:28.755  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:29.257  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:29.259  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:32.636  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:33.257  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
04-29 11:34:33.259  8175  8175 E ANDR-PERF: IPerf::tryGetService failed!
App version: 0.10.12-3328 (standard, 6aff438a, 59)
Android version: 9 (SDK 28)
Android build ID: 47.2.A.11.228
Device brand: Sony
Device manufacturer: Sony
Device name: G8341
Device model: G8341
Device product name: G8341```

**Comment by ghostbear** [comment]
Can you say a manga (with source) and what chapter this happens on, it would help to reproduce the bug 

Also, do you have "Always show chapter transition" on?

**Comment by Velorexe** [comment]
`Shikimori's Not Just a Cutie` from the `NineMangeEn` extension. I had it a lot around chapter 40-50 when I was reading yesterday, but now going back to it, Tachiyomi doesn't crash. Though it happened again this morning at around chapter 70. Unfortunately I can't reproduce the bug 100% of the time, but I'll check and upload more logs when they occur. My guess would be that it's an issue with the extension rather than Tachiyomi.

And yes, `Always show chapter transition` is on.

**Comment by Velorexe** [comment]
Crashed again at the end of chapter 79. Here's the crash log. I'll also check with other extensions and titles.

```
--------- beginning of crash
04-29 11:17:14.431  4992  4992 E AndroidRuntime: FATAL EXCEPTION: main
04-29 11:17:14.431  4992  4992 E AndroidRuntime: Process: eu.kanade.tachiyomi.debug, PID: 4992
04-29 11:17:14.431  4992  4992 E AndroidRuntime: java.lang.ArrayIndexOutOfBoundsException: length=31; index=-1
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:439)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewerAdapter.onPageSplit(PagerViewerAdapter.kt:141)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewer$onPageSplit$1.run(PagerViewer.kt:390)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:280)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6706)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-29 11:17:14.431  4992  4992 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
--------- beginning of main
04-29 15:37:20.251 14634 14643 E tachiyomi.debu: Something went wrong getting fds from adb. Retry!: Connection refused
04-29 15:37:23.523 14634 14634 E ANDR-PERF: IPerf::tryGetService failed!
04-29 15:37:27.680 14634 14634 E ANDR-PERF: IPerf::tryGetService failed!
04-29 15:37:32.340 14634 14634 E ANDR-PERF: IPerf::tryGetService failed!
04-29 15:37:33.632 14634 14634 E ANDR-PERF: IPerf::tryGetService failed!
04-29 15:37:33.656 14634 14634 E AndroidRuntime: FATAL EXCEPTION: main
04-29 15:37:33.656 14634 14634 E AndroidRuntime: Process: eu.kanade.tachiyomi.debug, PID: 14634
04-29 15:37:33.656 14634 14634 E AndroidRuntime: java.lang.ArrayIndexOutOfBoundsException: length=25; index=-1
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:439)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewerAdapter.onPageSplit(PagerViewerAdapter.kt:141)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewer$onPageSplit$1.run(PagerViewer.kt:390)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:280)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6706)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-29 15:37:33.656 14634 14634 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
04-29 15:37:33.662 14634 14634 E ACRA    : ACRA caught a ArrayIndexOutOfBoundsException for eu.kanade.tachiyomi.debug
04-29 15:37:33.662 14634 14634 E ACRA    : java.lang.ArrayIndexOutOfBoundsException: length=25; index=-1
04-29 15:37:33.662 14634 14634 E ACRA    : 	at java.util.ArrayList.get(ArrayList.java:439)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewerAdapter.onPageSplit(PagerViewerAdapter.kt:141)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at eu.kanade.tachiyomi.ui.reader.viewer.pager.PagerViewer$onPageSplit$1.run(PagerViewer.kt:390)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at android.os.Handler.handleCallback(Handler.java:873)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at android.os.Looper.loop(Looper.java:280)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at android.app.ActivityThread.main(ActivityThread.java:6706)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at java.lang.reflect.Method.invoke(Native Method)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-29 15:37:33.662 14634 14634 E ACRA    : 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
04-29 15:37:38.164 16198 16239 E SQLiteLog: (284) automatic index on C(manga_id)
04-29 15:37:38.164 16198 16239 E SQLiteLog: (284) automatic index on mangas_categories(manga_id)
04-29 15:37:41.251 16198 16198 E ANDR-PERF: IPerf::tryGetService failed!
04-29 15:37:41.254 16198 16198 E ANDR-PERF: IPerf::tryGetService failed!
App version: 0.10.12-3328 (standard, 6aff438a, 59)
Android version: 9 (SDK 28)
Android build ID: 47.2.A.11.228
Device brand: Sony
Device manufacturer: Sony
Device name: G8341
Device model: G8341
Device product name: G8341

**Comment by ghostbear** [comment]
I know it was related to `Dual Page Split`, and it was changed with 662b71436eea6ddda4d15d18ec0946e946c6e4e9 to work on IO instead of the main thread. The move to the IO thread seems to have introduced the bug.

And I was able to reproduce the bug with the manga and chapters you provided this helped to narrow down when this bug happened instead of using a brute force chapter.

I have found that this happens either when either of the two first pages of the next chapter or either of the two last pages of the previous chapter is a dual page spread.

I have solved this bug locally but it has introduced another one where the split page is gone
